b9lab-content/3-main-concepts/09-protobuf.md

```diff
@@ -45,18 +49,81 @@ The latest Google APIs will have gRPC versions of their interfaces, letting you
 * Reference to **codec**: Defaulted to go-amino, the codec in your Cosmos SDK application can be substituted with other suitable encoding frameworks as long as they persist data stores in byte slices and are deterministic.
 Reference to Module Manager: A reference to an object containing a list of the applications modules, known as the Module Manager.
 
-## Long-running exercise
-
-We have defined a number of data types: our transactions and our `FullGame`. Perhaps it is time to define them in Protobuf so as to enjoy its serialization facilities.
-
-We try to reuse the pre-existing `Game` if possible. Otherwise, we provide a transformer so as to reuse the rules code.
-
-<!-- TODO code detail. -->
+<ExpansionPanel title="Show me some code for my checkers blockchain">
+
+In the previous code samples, you saw something like:
+
+```go
+type StoredGame struct {
+    Creator string
+    Index string // The unique id that identifies this game.
+    Game string // The serialized board.
+    Turn string // "red" or "black"
+    Red string
+    Black string
+    Wager uint64
+}
+```
+With a _helpful_ note telling you that you still need to add serialization information like:
+
+```go
+type StoredGame struct {
+    Creator string `protobuf:"bytes,1,opt,name=creator,proto3" json:"creator,omitempty"`
+    ...
+}
+```
+How are you to add such cryptic commands without any error?
+
+## Move Upstream
+
+This is where Protobuf comes in to simplify your life yet more. The same `StoredGame` can in fact be declared as:
+
+```protobuf
+message StoredGame {
+  string creator = 1;
+  string index = 2;
+  string game = 3;
+  string turn = 4;
+  string red = 5;
+  string black = 6;
+  uint64 wager = 7;
+}
+```
+The `= 1` parts indicate how each field is identified in the serialized output, and additionally provides backward compatibility. Down the road, to preserve backward compatibility, if you add or remove fields, make sure to not reuse numbers for new fields, but to keep increasing the `= x` value.
+
+When _compiling_ it, Protobuf will add the `protobuf:"bytes..."` elements. Similarly, the messages to create a game can be declared in Protobuf as:
+
+```protobuf
+message MsgCreateGame {
+  string creator = 1;
+  string red = 2;
+  string black = 3;
+  uint64 wager = 4;
+}
+
+message MsgCreateGameResponse {
+  string idValue = 1;
+}
+```
+
+## Enter Starport
+
+Can this too be created for you? Yes. If you use Starport and commands like:
+
+```sh
+$ starport scaffold map storedGame game turn red black wager:uint --module checkers --no-message
+$ starport scaffold message createGame red black wager:uint --module checkers --response idValue
+```
+Head to [how to build your own chain](../5-my-own-chain/01-index) for more details on using Starport.
+
+When Starport creates a message for you, it also creates the gRPC definitions and Go handling code.
```

Comment:
Continued:

```suggestion

If you want to dive straight into coding your own chain, head to [how to build your own chain](../5-my-own-chain/01-index) for more details on using Starport.
```

Again, a closing summery & CTA before the expansion panel would be desirable.